contracts/javascore/centralized-connection/src/main/java/xcall/adapter/centralized/CentralizedConnection.java

```diff
@@ -0,0 +1,184 @@
+      * @return          The fee of sending a message to a given destination network
+      */
+     @External(readonly = true)
+     public BigInteger getFee(String to, boolean response) {
+         BigInteger messageFee = messageFees.get(to);
+         if (response) {
+             BigInteger responseFee = responseFees.get(to);
+             return messageFee.add(responseFee);
+         }
+         return messageFee;
+     }
+ 
+     /**
+      * Sends a message to the specified network.
+      *
+      * @param  to     Network Id of destination network 
+      * @param  svc    name of the service 
+      * @param  sn     positive for two-way message, zero for one-way message, negative for response
+      * @param  msg    serialized bytes of Service Message 
+      */
+     @Payable
+     @External
+     public void sendMessage(String to, String svc, BigInteger sn, byte[] msg) {
+         Context.require(Context.getCaller().equals(xCall.get()), "Only xCall can send messages");
+         BigInteger fee = BigInteger.ZERO;
+         if(sn.compareTo(BigInteger.ZERO)>=0) {
+             fee = getFee(to, true);
+         } else {
+            fee = getFee(to, false);
+         }
+
+         Context.require(Context.getValue().compareTo(fee) >= 0, "Insufficient balance");
+         Message(to, sn, msg);
+     }
+ 
+     /**
+      * Receives a message from a source network.
+      *
+      * @param  srcNetwork  the source network id from which the message is received
+      * @param  sn          the serial number of the message
+      * @param  msg         serialized bytes of Service Message 
+      */
+     @Payable
+     @External
+     public void recvMessage(String srcNetwork, BigInteger sn, byte[] msg) {
+        OnlyAdmin();
+        Context.require(!receipts.at(srcNetwork).getOrDefault(sn, false),"Duplicate Message");
+        receipts.at(srcNetwork).set(sn, true);
+        Context.call(xCall.get(), "handleMessage", srcNetwork, msg);
+     }
+
+     /**
+      * Reverts a message.
+      *
+      * @param  sn  the serial number of message representing the message to revert
+      */
+     @External
+     public void revertMessage(BigInteger sn) {
+        OnlyAdmin();
+        Context.call(xCall.get(), "handleError", sn);
+     }
+
+     /**
+      * Claim the fees.
+      *
+      */
+     @External
+     public void claimFees() {
+         OnlyAdmin();
+         Context.transfer(admin(), Context.getBalance(Context.getAddress()));
+     }
+
+     /**
+      * Get the receipts for a given source network and serial number.
+      *
+      * @param  srcNetwork  the source network id
+      * @param  sn          the serial number of message
+      * @return             the receipt if is has been recived or not
+      */
+     @External
+     public boolean getReceipts(String srcNetwork, BigInteger sn) {
+         OnlyAdmin();
+         return receipts.at(srcNetwork).getOrDefault(sn, false);
+     }
```

Comment:
I believe its read only method so no need to admin guard
```suggestion
     @External(readonly = true)
     public boolean getReceipts(String srcNetwork, BigInteger sn) {
         return receipts.at(srcNetwork).getOrDefault(sn, false);
     }
```

Reply:
Ohh, this was silly mistake 😅